### DISCUSSION TITLE
How to get the token certificate URL?

### DISCUSSION THREAD

**Discussion by XoDefender** [strongswan/strongswan#2260] [Q&A] [answered]
Is it possible to get the URL of a particular certificate? I am working over adding the gui for selecting token certificates and struggling with the selected cert identification.  Is it possible to get the url similar to **pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=20ac938ae45b6731;token=tokenemul1** and find a particular certificate which corresponds to the one I pass? 

I was trying to get it by **CKA_URL**, but **chunk_clone(chunk_create(attr[1].pValue, attr[1].ulValueLen));** was getting me null each time.

Thank you in advance for assistance!

**Comment by tobiasbrunner** [comment]
> Is it possible to get the URL of a particular certificate?

From where?

> I am working over adding the gui for selecting token certificates and struggling with the selected cert identification

What GUI are you referring to?

> I was trying to get it by **CKA_URL**

Get it from where? And if your token returns NULL, why are you asking here?

**Reply by XoDefender** [reply]
1) from the pkcs11 interface (pkcs11 plugin) in libstrongswan
2) I am referring to network-manager-strongswan GUI - I think that it is possible to add libnma cert chooser there and save the particular URL into some setting (ca-cert) => register it with NetworkManager => get from strongswan by **nm_setting_vpn_get_data_item(vpn, "ca-cert");** => find the particular certificate instance by the given URL (check all certificates registered in the pkcs11 set) => add it by **set_cert_and_key** in order to authenticate a client
3) The token is found by pkcs11 plugin, the certificate is found to, the problem is that the value of CKA_URL attribute in NULL, while label and value are correct ( 
**00[CFG]     loaded untrusted cert 'AstraCRT'
May 27 13:57:37 astra charon-systemd[23314]:     loaded untrusted cert 'AstraCRT'**):

CK_ATTRIBUTE attr[] = {
		{CKA_VALUE, NULL, 0},
		{CKA_LABEL, NULL, 0},
		{CKA_URL, NULL, 0},
		{CKA_TRUSTED, &trusted, sizeof(trusted)}
	};

entry = malloc(sizeof(*entry));
		entry->value = chunk_clone(
							chunk_create(attr[0].pValue, attr[0].ulValueLen));
		entry->label = chunk_clone(
							chunk_create(attr[1].pValue, attr[1].ulValueLen));
		entry->url = chunk_clone(
							chunk_create(attr[2].pValue, attr[2].ulValueLen));

**Reply by tobiasbrunner** [reply, marked as answer]
> 1. from the pkcs11 interface (pkcs11 plugin) in libstrongswan

What do you mean exactly?

> 2\. I am referring to network-manager-strongswan GUI - I think that it is possible to add libnma cert chooser there and save the particular URL into some setting (ca-cert) => register it with NetworkManager => get from strongswan by **nm_setting_vpn_get_data_item(vpn, "ca-cert");** => find the particular certificate instance by the given URL (check all certificates registered in the pkcs11 set) => add it by **set_cert_and_key** in order to authenticate a client

Maybe the URI can be translated to the lookup parameters used by strongSwan internally (module, slot, handle/ID)?

> 3\. , the problem is that the value of CKA_URL attribute in NULL

Why would that be necessary? That attribute is intended to store the URI to a certificate (maybe to fetch it from somewhere if it is not on the token, the PKCS#11 standard is not really specific for what it's actually used, though), not to match it against an URI or whatever you are trying to do.

**Reply by XoDefender** [reply]
That definitely makes sense. Is there a way to parse the URI in the strongswan code base, or it should be done manually? In libnma there is Gck module and gck_uri_parse function.

And I am not totally aware, but is there already a way to specify a particular certificate (on token) and use it for client authentication? As I know strongswan automatically selects the cert to use (https://docs.strongswan.org/docs/5.9/features/networkManager.html#_smart_card_requirements)

**Reply by tobiasbrunner** [reply]
> Is there a way to parse the URI in the strongswan code base, or it should be done manually? In libnma there is Gck module and gck_uri_parse function.

Don't know. Maybe in the GUI itself depending on what was selected with the certificate chooser (whose documentation seems to be completely offline).

> And I am not totally aware, but is there already a way to specify a particular certificate (on token) and use it for client authentication?

Not in the NM plugin, but for many use cases that's also not necessary. Configuration of the regular daemon via swanctl/vici allows selecting specific certificates, though.

**Reply by XoDefender** [reply]
Does it mean that the general algorithm for selecting a certificate may be  overridden:
(https://docs.strongswan.org/docs/5.9/features/networkManager.html#_smart_card_requirements)

And the one for which I specify data (handle, slot) in the swanctl.conf will be chosen instead (for a particular connection)?

**Reply by tobiasbrunner** [reply]
That algorithm is only used by the NetworkManager plugin, the regular daemon does not use that.

**Reply by XoDefender** [reply]
Thanks! I am getting a little bit closer to understanding the way it works. It is not relevant to this particular topic, but I am interested in the reason, why in time I call:

char* crt_key = "67 86 8D 7F 53 41 4D 8B 8F 73 97 05 DD B9 EB E4 8F 23 6A 0B";
chunk_t crt_keyid = chunk_from_hex(chunk_create(keyid, strlen(keyid)), NULL);
certificate_t *token_cert = lib->creds->create(lib->creds, CRED_CERTIFICATE,
								KEY_ANY, BUILD_PKCS11_KEYID, crt_keyid, BUILD_END); - in (**find_smartcard_key: nm-service.c**)

The **enumerator->enumerate(enumerator, &p11, &current)** in (**pkcs11_creds_load: pkcs11_creds.c**) returns NULL each time I try to call the pkcs11 builder from **nm_service.c** (lib->creds->create ...)? It does not even go through the first token slot. When I use the same code not in the charon-nm module, but in the (**find_certificates: pkcs11_creds.c**) it works correctly and the enumerator is not NULL.

Could you please give me a hint why it could theoretically happen? Is it because something is wrong in config files, or it is a code thing?

**Reply by tobiasbrunner** [reply]
First, that ID won't be correct. Don't add spaces between bytes (the only separator supported by `chunk_from_hex()` is `:`).

Second, you might not have configured the PKCS#11 module correctly. `charon` and `charon-nm` are configured separately (that's different from e.g. `charon-systemd` who inherits the `charon` settings). So you either have to configure the module in `charon-nm.plugins.pkcs11.modules` or in `libstrongswan.plugins.pkcs11.modules` (the latter configures the modules for all daemons).

**Reply by XoDefender** [reply]
the entire strongswan.conf looks this way, seems to be correct for the pkcs modules loading:
[strongswan.txt](https://github.com/strongswan/strongswan/files/15471174/strongswan.txt)

Plugins seem to be loaded correctly to: 
May 28 17:32:53 charon-nm[2911]: 00[LIB] loaded plugins: nm-backend charon-nm pkcs11 aes des rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pkcs1 pkcs7 pkcs8 sshkey pem fips-prf gmp curve25519 xcbc cmac hmac drbg kernel-netlink socket-default

If I comment the token enumerator code and reach:
**rv = p11->f->C_OpenSession(current, CKF_SERIAL_SESSION, NULL, &session);**

The charon-nm crashes with: **charon-nm[7687]: 05[DMN] killing ourself, received critical signal**

Errors at the OpenSession execution are:
May 28 18:23:40  charon-nm[7687]: 05[DMN] thread 5 received 11
May 28 18:23:40  NetworkManager[956]:  dumping 24 stack frame addresses:
May 28 18:23:40  NetworkManager[956]:   /lib/x86_64-linux-gnu/libpthread.so.0 @ 0x7ee1440ca000 [0x7ee1440dc730]
May 28 18:23:40  NetworkManager[956]:     -> ??:?

**Reply by XoDefender** [reply]
Yes, it seems that smth is wrong with accessing the library:

**this->libs->get_count(this->libs);** is 0 (create_token_enumerator: pkcs11_manager.c)
